tools/lemmacmd.go

```diff
@@ -0,0 +1,267 @@
+
+	// derive key and return it
+	key, err = deriveKeyFromPassphraseAndSalt(salt, keyiter, passphrase)
+	if err != nil {
+		return nil, fmt.Errorf("unable to derive key from passphrase: %v", err)
+	}
+
+	return key, nil
+}
+
+// Given a passphrase, generates a 128-bit salt and uses those as input to PBKDF#2 to derive 32-byte key.
+func deriveKeyFromPassphrase(keyiter int, passphrase string) (salt []byte, key *[secret.SecretKeyLength]byte, err error) {
+	// generate 128-bit salt
+	csprng := random.CSPRNG{}
+	salt, err = csprng.Bytes(16)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// derive key and convert to array
+	keySlice := pbkdf2.Key([]byte(passphrase), salt, keyiter, 32, sha256.New)
+	keyBytes, err := secret.KeySliceToArray(keySlice)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// derive key from salt and passphrase
+	return salt, keyBytes, nil
+}
+
+// Given a passphrase and a salt, uses those as input to PBKDF#2 to derive 32-byte key.
+func deriveKeyFromPassphraseAndSalt(salt []byte, keyiter int, passphrase string) (*[secret.SecretKeyLength]byte, error) {
+	// derive key and convert to array
+	keySlice := pbkdf2.Key([]byte(passphrase), salt, keyiter, 32, sha256.New)
+	keyBytes, err := secret.KeySliceToArray(keySlice)
+	if err != nil {
+		return nil, err
+	}
+
+	// derive key from salt and passphrase
+	return keyBytes, nil
+}
+
+// Encodes all data needed to decrypt message into a JSON string and writes it to disk.
+func writeCiphertext(salt []byte, keyiter int, sealed *secret.SealedBytes, filename string) error {
+	// salt, nonce, and ciphertext are stored base64 encoded
+	ec := EncodedCiphertext{
+		KeySalt:         base64.StdEncoding.EncodeToString(salt),
+		KeyIter:         keyiter,
+		CiphertextNonce: base64.StdEncoding.EncodeToString(sealed.Nonce),
+		Ciphertext:      base64.StdEncoding.EncodeToString(sealed.Ciphertext),
+	}
+
+	// marshal encoded ciphertext into a json string
+	b, err := json.MarshalIndent(ec, "", "   ")
+	if err != nil {
+		return err
+	}
+
+	// write to disk with read only permissions for the current user
+	err = ioutil.WriteFile(filename, b, 0600)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Reads in encoded ciphertext from disk and breaks into component parts.
+func readCiphertext(filename string) ([]byte, int, *secret.SealedBytes, error) {
+	var err error
+
+	plaintextBytes, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, 0, nil, err
+	}
+
+	var ec EncodedCiphertext
+	err = json.Unmarshal(plaintextBytes, &ec)
+	if err != nil {
+		return nil, 0, nil, err
+	}
+
+	var saltBytes []byte
+	if ec.KeySalt != "" {
+		saltBytes, err = base64.StdEncoding.DecodeString(ec.KeySalt)
+		if err != nil {
+			return nil, 0, nil, err
+		}
+	}
+
+	nonceBytes, err := base64.StdEncoding.DecodeString(ec.CiphertextNonce)
+	if err != nil {
+		return nil, 0, nil, err
+	}
+
+	cipherhtextBytes, err := base64.StdEncoding.DecodeString(ec.Ciphertext)
+	if err != nil {
+		return nil, 0, nil, err
+	}
```

Comment:
You should make `KeySalt`, `CiphertextNonce`, and `Ciphertext` `[]byte`s instead of `string`s. If you do, the json package will take care of base64 encoding and decoding for you.

> Array and slice values encode as JSON arrays, except that []byte encodes as a base64-encoded string, and a nil slice encodes as the null JSON object.

http://golang.org/pkg/encoding/json/#Marshal